### PR TITLE
Settings: Metric Configuration: Refactor & Clean Up (5/n)

### LINKED ISSUE #81: [Frontend] New Settings Page: Metric Configuration - Refactor & Clean Up (5/n)

### PR DESCRIPTION
**Pull request by mxosman**
## Description of the change

Cleans up the Metric Configuration component and specifically accomplishes the following items:
- Move components into their respective files and general code clean up
- Merge newly created types for metric config with original types
- Fix bugs (switching agencies, out-of-sync checking and unchecking)
- Handles case where no definitions are sent

## Related issues

Closes #81 

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/components/MetricConfiguration/Configuration.tsx [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    typeOfUpdate: "METRIC" | "DISAGGREGATION" | "DIMENSION" | "CONTEXT",
+    updatedSetting: MetricSettings,
+    debounce?: boolean
+  ) => void;
+  setActiveDimension: React.Dispatch<
+    React.SetStateAction<MetricDisaggregationDimensions | undefined>
+  >;
+};
+
+export const Configuration: React.FC<MetricConfigurationProps> = ({
```
Comment (mxosman) [inline publisher/src/components/MetricConfiguration/Configuration.tsx:57]: No changes to this component except for a bug fix in L91. Would love to know if naming this file `Configuration` is confusing or if there is a better name to specify this as the actual configuration portion of the `MetricConfiguration` component. 😅

File: publisher/src/components/MetricConfiguration/ContextConfiguration.tsx [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+              >
+                Clear Input
+              </BinaryRadioGroupClearButton>
+            </BinaryRadioGroupContainer>
+          )}
+        </MetricContextItem>
+      ))}
+    </MetricContextContainer>
+  );
+};
```
Comment (mxosman) [inline publisher/src/components/MetricConfiguration/ContextConfiguration.tsx:218]: No changes here aside from updating types/names.

File: publisher/src/components/MetricConfiguration/MetricBox.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      <MetricName>{displayName}</MetricName>
+      <MetricDescription>{description}</MetricDescription>
+      <MetricNameBadgeWrapper>
+        <Badge color={!enabled ? "GREY" : "GREEN"} disabled={!enabled} noMargin>
+          {!enabled ? "Inactive" : frequency.toLowerCase()}
+        </Badge>
+      </MetricNameBadgeWrapper>
+    </MetricBoxContainer>
+  );
+};
```
Comment (mxosman) [inline publisher/src/components/MetricConfiguration/MetricBox.tsx:60]: No changes.

File: publisher/src/components/MetricConfiguration/MetricDefinitions.tsx [hunk 4]
```diff
@@ -130,52 +130,70 @@ export const MetricDefinitions: React.FC<MetricDefinitionsProps> = ({
...
+            </Definitions>
+          </>
+        )}
+
+        {/* Display when user is viewing a dimension & there are no settings available */}
+        {!mockDefinitionsToDisplay?.settings?.length && activeDimension && (
+          <DefinitionsSubTitle>
+            This breakdown has no customizations available yet.
+          </DefinitionsSubTitle>
+        )}
```
Comment (mxosman) [inline publisher/src/components/MetricConfiguration/MetricDefinitions.tsx:184]: No major changes here aside from handling the case where the settings object is not present/is empty.

Here's how this looks for a top-level metric definition and a dimension definition:

https://user-images.githubusercontent.com/59492998/196734599-102d172e-bad2-47e2-b180-1324c999443e.mov

**Review by terryttsai — APPROVED** [review]
Overall structure looks good to me! Part of me wonders what a mobx store for fetching and storing the metric settings would look like. 

In redux we could have the metric settings as well as the active metric/disaggregation/dimension live in the redux state and then we wouldn't need to pass all these states and state-setters as props between all these nicely-broken up components...

Though without a central store this still looks good to me and I could follow along as I stepped through this

**Comment by mxosman** [comment]
> Overall structure looks good to me! Part of me wonders what a mobx store for fetching and storing the metric settings would look like.
> 
> In redux we could have the metric settings as well as the active metric/disaggregation/dimension live in the redux state and then we wouldn't need to pass all these states and state-setters as props between all these nicely-broken up components...
> 
> Though without a central store this still looks good to me and I could follow along as I stepped through this

Thank you for the feedback, Terry! Yeah, I actually went back and forth on whether or not to create a store and ultimately decided not to go that deep. I'm still on the fence - I might change my mind in a subsequent PR, but for now I think this at the very least adds some clarity to the feature and will be a good place to implement a store from if we decide it's worth doing.

**Comment by terryttsai** [comment]
Yeah that reasoning sounds good to me. Thanks for taking the time to refactor and structure things out in such an organized way!

**Comment by lilidworkin** [comment]
Nice, thanks for cleaning things up @mxosman !

**Pull request merged**